### ISSUE TITLE
segfault with 1.0.0

### ISSUE THREAD

**Issue by mlilien** [Azure/iot-hub-device-update#269]
the following log includes a segfault after ~18h uptime
azure-iot-sdk-c version:  LTS_07_2022_Ref02

log:
```
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 04:41:05 PINGREQ
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 04:41:05 PINGRESP
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 04:45:05 PINGREQ
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 04:45:05 PINGRESP
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 04:49:05 PINGREQ
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 04:49:05 PINGRESP
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 04:53:05 PINGREQ
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 04:53:05 PINGRESP
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 04:57:05 PINGREQ
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 04:57:05 PINGRESP
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:01:05 PINGREQ
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:01:05 PINGRESP
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:05:05 PINGREQ
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:05:06 PINGRESP
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:09:06 PINGREQ
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:09:06 PINGRESP
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:13:06 PINGREQ
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:13:06 PINGRESP
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:17:06 PINGREQ
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:17:06 PINGRESP
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:21:06 PINGREQ
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:21:06 PINGRESP
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:25:05 DISCONNECT
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T05:25:05.8709Z 50444[50444] [D] IotHub connection status: 1, reason: 0 [ADUC_ConnectionStatus_Callback]
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T05:25:05.8710Z 50444[50444] [W] IotHub connection SAS token expired. Attempting refresh. [ADUC_ConnectionStatus_Callback]
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T05:25:05.8712Z 50444[50444] [I] Requesting connection string from the Edge Identity Service [GetConnectionInfoFromIdentityService]
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T05:25:08.1059Z 50444[50444] [I] Attempting to create connection to IotHub using type: ADUC_ConnType_Module  [ADUC_DeviceClient_Create]
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T05:25:08.1061Z 50444[50444] [I] IotHub Protocol: MQTT [GetIotHubProtocolFromConfig]
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T05:25:08.1062Z 50444[50444] [I] IoTHub Device Twin callback registered. [ADUC_DeviceClient_Create]
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T05:25:08.1062Z 50444[50444] [I] Refreshing the handle for the PnP channels. [ADUC_PnP_Components_HandleRefresh]
Dec 01 05:25:08 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T05:25:08.1062Z 50444[50444] [I] Successfully refreshed SAS Token [ADUC_Refresh_IotHub_Connection_SAS_Token]
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:25:08 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 192 | USERNAME: XXX.azure-devices.net/YYY/AducIotAgent/?api-version=2020-09-30&DeviceClientType=iothubclient%2f1.9.1%20(native%3b%20Linux%3b%20aarch64)&model-id=dtmi%3aazure%3aiot%3adeviceUpdateModel%3b2 | PWD: XXXX | CLEAN: 0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:25:10 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T05:25:10.6778Z 50444[50444] [D] IotHub connection status: 0, reason: 6 [ADUC_ConnectionStatus_Callback]
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:25:10 SUBSCRIBE | PACKET_ID: 2 | TOPIC_NAME: $iothub/twin/res/# | QOS: 0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:25:10 SUBACK | PACKET_ID: 2 | RETURN_CODE: 0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:25:10 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/GET/?$rid=3
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:25:11 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/res/200/?$rid=3 | PAYLOAD_LEN: 707
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:25:11 SUBSCRIBE | PACKET_ID: 4 | TOPIC_NAME: $iothub/twin/PATCH/properties/desired/# | QOS: 0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:25:11 SUBACK | PACKET_ID: 4 | RETURN_CODE: 0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:29:11 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:29:11 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:33:11 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:33:11 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:37:11 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:37:11 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:41:11 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:41:11 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:45:11 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:45:11 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:49:11 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:49:11 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:53:11 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:53:11 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 05:57:11 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 05:57:11 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:01:11 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:01:11 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:05:11 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:05:11 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:09:11 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:09:11 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:13:09 DISCONNECT
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T06:13:09.3299Z 50444[50444] [D] IotHub connection status: 1, reason: 0 [ADUC_ConnectionStatus_Callback]
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T06:13:09.3299Z 50444[50444] [W] IotHub connection SAS token expired. Attempting refresh. [ADUC_ConnectionStatus_Callback]
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T06:13:09.3302Z 50444[50444] [I] Requesting connection string from the Edge Identity Service [GetConnectionInfoFromIdentityService]
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T06:13:12.0073Z 50444[50444] [I] Attempting to create connection to IotHub using type: ADUC_ConnType_Module  [ADUC_DeviceClient_Create]
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T06:13:12.0074Z 50444[50444] [I] IotHub Protocol: MQTT [GetIotHubProtocolFromConfig]
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T06:13:12.0075Z 50444[50444] [I] IoTHub Device Twin callback registered. [ADUC_DeviceClient_Create]
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T06:13:12.0075Z 50444[50444] [I] Refreshing the handle for the PnP channels. [ADUC_PnP_Components_HandleRefresh]
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T06:13:12.0075Z 50444[50444] [I] Successfully refreshed SAS Token [ADUC_Refresh_IotHub_Connection_SAS_Token]
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:13:12 CONNECT | VER: 4 | KEEPALIVE: 240 | FLAGS: 192 | USERNAME: XXX.azure-devices.net/YYY/AducIotAgent/?api-version=2020-09-30&DeviceClientType=iothubclient%2f1.9.1%20(native%3b%20Linux%3b%20aarch64)&model-id=dtmi%3aazure%3aiot%3adeviceUpdateModel%3b2 | PWD: XXXX | CLEAN: 0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:13:12 CONNACK | SESSION_PRESENT: true | RETURN_CODE: 0x0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T06:13:12.4617Z 50444[50444] [D] IotHub connection status: 0, reason: 6 [ADUC_ConnectionStatus_Callback]
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:13:12 SUBSCRIBE | PACKET_ID: 2 | TOPIC_NAME: $iothub/twin/res/# | QOS: 0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:13:12 SUBACK | PACKET_ID: 2 | RETURN_CODE: 0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:13:12 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/GET/?$rid=3
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:13:12 PUBLISH | IS_DUP: false | RETAIN: 0 | QOS: DELIVER_AT_MOST_ONCE | TOPIC_NAME: $iothub/twin/res/200/?$rid=3 | PAYLOAD_LEN: 707
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:13:12 SUBSCRIBE | PACKET_ID: 4 | TOPIC_NAME: $iothub/twin/PATCH/properties/desired/# | QOS: 0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:13:13 SUBACK | PACKET_ID: 4 | RETURN_CODE: 0
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:17:13 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:17:13 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:21:13 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:21:13 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:25:13 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:25:13 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:29:13 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:29:13 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:33:13 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:33:13 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:37:13 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:37:13 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:41:13 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:41:13 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:45:13 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:45:13 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:49:13 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:49:13 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:53:13 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:53:13 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 06:57:13 PINGREQ
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: <- 06:57:13 PINGRESP
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: -> 07:01:13 DISCONNECT
Dec 01 07:01:13 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T07:01:13.2407Z 50444[50444] [D] IotHub connection status: 1, reason: 0 [ADUC_ConnectionStatus_Callback]
Dec 01 07:49:17 phygate-tauri-l-imx8mm-2 AducIotAgent[50444]: 2022-12-01T07:01:13.2407Z
Dec 01 07:49:17 phygate-tauri-l-imx8mm-2 systemd[1]: deviceupdate-agent.service: Main process exited, code=dumped, status=11/SEGV
Dec 01 07:49:17 phygate-tauri-l-imx8mm-2 systemd[1]: deviceupdate-agent.service: Failed with result 'core-dump'.
```
backtrace: 
```
Core was generated by `/usr/bin/AducIotAgent -l 0 -e'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000aaaae7b10028 in mqtt_client_dowork (handle=0x642d6572757a612e) at /usr/src/debug/azure-iot-sdk-c/AUTOINC+3fd808bee3-r0/git/umqtt/src/mqtt_client.c:1379
1379    /usr/src/debug/azure-iot-sdk-c/AUTOINC+3fd808bee3-r0/git/umqtt/src/mqtt_client.c: No such file or directory.
[Current thread is 1 (LWP 50444)]
(gdb) bt
#0  0x0000aaaae7b10028 in mqtt_client_dowork (handle=0x642d6572757a612e) at /usr/src/debug/azure-iot-sdk-c/AUTOINC+3fd808bee3-r0/git/umqtt/src/mqtt_client.c:1379
#1  0x0000aaaae7adb018 in IoTHubTransport_MQTT_Common_DoWork (handle=0xaaab07d09f50) at /usr/src/debug/azure-iot-sdk-c/AUTOINC+3fd808bee3-r0/git/iothub_client/src/iothubtransport_mqtt_common.c:3643
#2  0x0000aaaae7ac1258 in main (argc=<optimized out>, argv=<optimized out>) at /usr/src/debug/iot-hub-device-update/AUTOINC+33554d2947-r0/git/src/agent/src/main.c:1556
(gdb) i threads
  Id   Target Id         Frame 
* 1    LWP 50444         0x0000aaaae7b10028 in mqtt_client_dowork (handle=0x642d6572757a612e) at /usr/src/debug/azure-iot-sdk-c/AUTOINC+3fd808bee3-r0/git/umqtt/src/mqtt_client.c:1379
  2    LWP 50445         0x0000ffff89aa4470 in __GI___clock_nanosleep (clock_id=<optimized out>, clock_id@entry=0, flags=flags@entry=0, req=req@entry=0xffff8969dd30, rem=rem@entry=0xffff8969dd30) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:48
  3    LWP 50448         0x0000ffff89ac9ff4 in __libc_open64 (file=file@entry=0xaaaae7bda3e0 "/var/lib/adu/du-commands.fifo", oflag=oflag@entry=0) at ../sysdeps/unix/sysv/linux/open64.c:41
(gdb) 
```

**Comment by eshashah-msft** [comment]
@mlilien Thanks for reporting this. We are tracking this bug and will make a fix available soon.

**Comment by mlilien** [comment]
additional information: we patched the 1.0.0 version in the above scenario with #243, because otherwise we weren't able to update at all.

**Comment by Nox-MSFT** [comment]
Hi @mlilien , this fix is included in [1.0.1 release](https://github.com/Azure/iot-hub-device-update/releases). Please give it a try. Thank you.
